### PR TITLE
Environment_oM: Added new objects for time dependent analysis grid results

### LINKED ISSUE #1198: Environment_oM: Create objects to enable analysis mesh with timestep result workflows
#### Description:
Per workshop 2/17/21 for managing analysis results for meshes and analysis grids, we need to create a set of result objects to streamline handling results at nodes across a grid over a stretch of time (ie hourly for a year). 

Possible structure as discussed on the call:
General result object with result value(s) per analyses and timesteps with datetime included
Nodes each contain a Point location and Result object above
Analysis Grid containing nodes

Relates to #1189 

Test cases for this include combined daylighting and artificial lighting analysis, external microclimate analysis (many points with various values per different metrics / analyses), raw data including datetime and value for POE, and space level results translated from IES.

### PR DESCRIPTION
**Pull request by JohannaOlin1**
### Issues addressed by this PR

Closes #1198 

Added TimeIndexResult object and IAnalysisResult object. Also added TimeIndexResult as property of Node object. 
Drawn workflow of the new objects here: [Time dependent analysis results.pptx](https://github.com/BHoM/BHoM/files/6035257/Time.dependent.analysis.results.pptx)

### CODE REVIEW

**Comment by tg359** [comment]
I'm going to be adding some unit-specific `AnalysisResult` objects shortly. These will probably look like this:

```c#
public class IlluminanceResult: BHoMObject, IAnalysisResult
{
    public virtual double Result { get; set; } = new double();
    public virtual Unit Unit { get; } = Unit.lux;
}

public class TemperatureResult: BHoMObject, IAnalysisResult
{
    public virtual double Result { get; set; } = new double();
    public virtual Unit Unit { get; } = Unit.celsius;
}
```

You might notice a `Unit` enum being used there. Is there one of these floating around in BHoM-land that I could make use of? I've not used any of the localisation stuff yet properly so I'd be surprised if not.

Also, @FraserGreenroyd & @enarhi, why is `Result` a single value, and not a list? We could still use it for single values, but we'd also be able to expand if if needed where we had multiple `IAnalysisResults` assigned to a `TimeIndexResult`.

**Comment by tg359** [comment]
I added some results objects linked to the `Quantities_oM` to assign units. They should work, but Create isn't working in GH for now. For example, try adding a `DryBulbTemperature` object to a canvas and it states it's not linked to a method.

**Comment by tg359** [comment]
Related branch here - [BHoM_Engine#2371](https://github.com/BHoM/BHoM_Engine/tree/Environment_Engine-%232371-IAnalysisResultOperations)

[BHoM#1198_BHoMEngine#2371.zip](https://github.com/BHoM/BHoM/files/6103083/BHoM.1198_BHoMEngine.2371.zip)

**Comment by tg359** [comment]
@BHoMBot check required

**Comment by tg359** [comment]
@BHoMBot check required

**Comment by tg359** [comment]
@BHoMBot check required

**Comment by tg359** [comment]
@enarhi @FraserGreenroyd 
As Johanna opening this PR I can't re-raise without closing and creating a new one, but pending the BHoMBot checks, could you give this a review?

There's an associated issue raised on Engine about the creation and use of these in association with `AnalysisGrids` https://github.com/BHoM/BHoM_Engine/issues/2512.

**Review by jamesramsden-bh — COMMENTED** [review]
I have performed a general review as per the task on the Priority Workflow board. I successfully compiled the code and was able to run the example gh file provided by @tg359 on Issue #1198. 
Without a test procedure in place (correct me if I'm wrong?) I believe somebody other than myself who has more knowledge of the functionality should provide the approving review, but they may use this comment to support that approving review if appropriate.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check serialisation

File: Environment_oM/Results/ResultObjects/Angle/WindDirectionResult.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing wind-direction data.")]
+    public class WindDirectionResult : BHoMObject, IAngleResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Angle]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Angle/WindDirectionResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/EnergyPerUnitArea/DiffuseHorizontalRadiationResult.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    [Description("A results object containing diffuse horizontal radiation data.")]
+    public class DiffuseHorizontalRadiationResult : BHoMObject, IEnergyPerUnitAreaResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [EnergyPerUnitArea]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/EnergyPerUnitArea/DiffuseHorizontalRadiationResult.cs:39]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/EnergyPerUnitArea/DirectNormalRadiationResult.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing direct normal radiation data.")]
+    public class DirectNormalRadiationResult : BHoMObject, IEnergyPerUnitAreaResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [EnergyPerUnitArea]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/EnergyPerUnitArea/DirectNormalRadiationResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/EnergyPerUnitArea/GlobalHorizontalRadiationResult.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    [Description("A results object containing global horizontal radiation data.")]
+    public class GlobalHorizontalRadiationResult : BHoMObject, IEnergyPerUnitAreaResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [EnergyPerUnitArea]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/EnergyPerUnitArea/GlobalHorizontalRadiationResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/EnergyPerUnitArea/InsolationResult.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing insolation data.")]
+    public class InsolationResult : BHoMObject, IEnergyPerUnitAreaResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [EnergyPerUnitArea]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/EnergyPerUnitArea/InsolationResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/EnergyPerUnitTime/PowerResult.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing power data.")]
+    public class PowerResult : BHoMObject, IEnergyPerUnitTimeResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [EnergyPerUnitTime]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/EnergyPerUnitTime/PowerResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Illuminance/IlluminanceResult.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    [Description("A results object containing illuminance data.")]
+    public class IlluminanceResult : BHoMObject, IAnalysisResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Quantities.Attributes.Illuminance] // TODO - for some reason not including traceback to Quantities results in error here
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Illuminance/IlluminanceResult.cs:37]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Pressure/AtmosphericPressureResult.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing atmospheric pressure data.")]
+    public class AtmosphericPressureResult : BHoMObject, IPressureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Pressure]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Pressure/AtmosphericPressureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Ratio/HumidityRatioResult.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    [Description("A results object containing humidity ratio data.")]
+    public class HumidityRatioResult : BHoMObject, IRatioResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Ratio]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Ratio/HumidityRatioResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Ratio/RelativeHumidityResult.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing relative humidity data.")]
+    public class RelativeHumidityResult : BHoMObject, IRatioResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Ratio]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Ratio/RelativeHumidityResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/SpecificEnergy/EnthalpyResult.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    [Description("A results object containing enthalpy data.")]
+    public class EnthalpyResult : BHoMObject, ISpecificEnergyResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [SpecificEnergy]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/SpecificEnergy/EnthalpyResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/DewPointTemperatureResult.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing dew point temperature data.")]
+    public class DewPointTemperatureResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/DewPointTemperatureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/DryBulbTemperatureResult.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    [Description("A results object containing dry bulb temperature data.")]
+    public class DryBulbTemperatureResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/DryBulbTemperatureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/MeanRadiantTemperatureResult.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing mean radiant temperature data.")]
+    public class MeanRadiantTemperatureResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/MeanRadiantTemperatureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/SkyTemperatureResult.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing sky temperature data.")]
+    public class SkyTemperatureResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/SkyTemperatureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/SurfaceTemperatureResult.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing surface temperature data.")]
+    public class SurfaceTemperatureResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/SurfaceTemperatureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/UniversalThermalClimateIndexResult.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing Universal Thermal Climate Index data.")]
+    public class UniversalThermalClimateIndexResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/UniversalThermalClimateIndexResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/Temperature/WetBulbTemperatureResult.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing wet bulb temperature data.")]
+    public class WetBulbTemperatureResult : BHoMObject, ITemperatureResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Temperature]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Temperature/WetBulbTemperatureResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

File: Environment_oM/Results/ResultObjects/TimeIndexResult.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+using System.ComponentModel;
+
+namespace BH.oM.Environment.Results
+{
+    [Description("A collection of results associated with a point-in-time.")]
+    public class TimeIndexResult : BHoMObject
+    {
+        public virtual List<IAnalysisResult> Results { get; set; } = new List<IAnalysisResult>();
+
+        public virtual int TimeIndex { get; set; } = new int();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/TimeIndexResult.cs:34]:
```suggestion
        public virtual int TimeIndex { get; set; } = 0;
```

File: Environment_oM/Results/ResultObjects/Velocity/WindSpeedResult.cs [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    [Description("A results object containing wind speed data")]
+    public class WindSpeedResult : BHoMObject, IVelocityResult
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Velocity]
+        [Description("The value associated with the results object.")]
+        public virtual double Result { get; set; } = new double();
```
Comment (FraserGreenroyd) [inline Environment_oM/Results/ResultObjects/Velocity/WindSpeedResult.cs:38]:
```suggestion
        public virtual double Result { get; set; } = 0.0;
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

**Review by FraserGreenroyd — APPROVED** [review]
This now all LGTM, great work @JohannaOlin1 and @tg359 on getting this over the line :smile:

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**